android/src/main/java/com/reactnativeimagecolors/ImageColorsModule.kt

```diff
@@ -0,0 +1,162 @@
+package com.reactnativeimagecolors
+
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.graphics.Color
+import android.util.Base64
+import android.webkit.URLUtil
+import androidx.annotation.NonNull
+import androidx.palette.graphics.Palette
+
+import expo.modules.kotlin.Promise
+import expo.modules.kotlin.modules.Module
+import expo.modules.kotlin.modules.ModuleDefinition
+import expo.modules.kotlin.records.Field
+import expo.modules.kotlin.records.Record
+
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+
+import java.net.URI
+
+class Config : Record {
+  @Field
+  val fallback: String = "#000000"
+
+  @Field
+  val headers: Map<String, String>? = null
+
+  @Field
+  val pixelSpacing: Int = 5
+}
+
+class ImageColorsModule : Module() {
+
+  /**
+   * https://gist.github.com/maxjvh/a6ab15cbba9c82a5065d
+   * pixelSpacing tells how many pixels to skip each pixel.
+   * If pixelSpacing > 1: the average color is an estimate, but higher values mean better performance.
+   * If pixelSpacing == 1: the average color will be the real average.
+   * If pixelSpacing < 1: the method will most likely crash (don't use values below 1).
+   */
+  private fun calculateAverageColor(@NonNull bitmap: Bitmap, pixelSpacing: Int): Int {
+    var R = 0
+    var G = 0
+    var B = 0
+
+    val height: Int = bitmap.height
+    val width: Int = bitmap.width
+
+    var n = 0
+    val pixels = IntArray(width * height)
+
+    bitmap.getPixels(pixels, 0, width, 0, 0, width, height)
+
+    var i = 0
+
+    while (i < pixels.size) {
+      val color = pixels[i]
+      R += Color.red(color)
+      G += Color.green(color)
+      B += Color.blue(color)
+      n++
+      i += pixelSpacing
+    }
+    return Color.rgb(R / n, G / n, B / n)
+  }
+
+  private fun getHex(rgb: Int): String {
+    return String.format("#%06X", 0xFFFFFF and rgb)
+  }
+
+  private fun handleError(promise: Promise, err: Exception) {
+    GlobalScope.launch(Dispatchers.Main) {
+      promise.reject("[ImageColors] Error", err.message, err)
+    }
+  }
+
+  override fun definition() = ModuleDefinition {
+    Name("ImageColors")
+
+    AsyncFunction("getColors") { uri: String, config: Config, promise: Promise ->
+      try {
```

Comment:
`runOnQueue()` does not allow to create special queue, so i restore thread pool with `CoroutineScope` from kotlin

Reply:
When I checked the example from #23 I encountered the error `Android:java.lang.OutOfMemoryError` and the solution was to add `android:largeHeap="true"` to `AndroidManifest.xml` in example app. Should we mention this in the documentation ?

Reply:
interesting, will check it out and get back to you

Reply:
could you share how exactly did you run into that error? I used the dev build for the example app in this PR and tried calling `getColors` 7 times with different images in `Promise.all([...])` on a low-end Android. I was profiling the memory and didn't run into the issue 🤔

Reply:
I used [image from issue example](https://i.ibb.co/RzGLN8P/edgar-castrejon-1-SPu0-KT-Ejg-unsplash.jpg) (2.7mb). I am using emulator sice i dont have real android device. This is full error
```
java.lang.OutOfMemoryError: Failed to allocate a 70353768 byte allocation with 6291456 free bytes and 32MB until OOM, target footprint 22207576, growth limit 50331648
        at com.reactnativeimagecolors.ImageColorsModule.calculateAverageColor(ImageColorsModule.kt:54)
        at com.reactnativeimagecolors.ImageColorsModule.access$calculateAverageColor(ImageColorsModule.kt:35)
```

It looks like the error occurs when calling `calculateAverageColor`. I debug it and it crash on [`IntArray`](https://github.com/KrzysztofMoch/react-native-image-colors/blob/20d4e869e670d42b8a3ef45d4f08492fe66a4f5a/android/src/main/java/com/reactnativeimagecolors/ImageColorsModule.kt#L54). 

> docs say `IntArray` -  Creates a new array of the specified size, with all elements initialized to zero.

If I understand correctly, this function generates height * width array.  So for this example it generate 3751 × 4689 array, 17 588 439 elements ?

Reply:
I created new implementation for `calculateAverageColor`
```kotlin
private fun calculateAverageColor(bitmap: Bitmap, pixelSpacing: Int): Int {
    var redSum = 0
    var greenSum = 0
    var blueSum = 0
    var pixelCount = 0

    for (x in 0 until bitmap.width step pixelSpacing) {
      for (y in 0 until bitmap.height step pixelSpacing) {
        val pixel = bitmap.getPixel(x, y)
        redSum += Color.red(pixel)
        greenSum += Color.green(pixel)
        blueSum += Color.blue(pixel)
        pixelCount++
      }
    }

    val redAvg = redSum / pixelCount
    val greenAvg = greenSum / pixelCount
    val blueAvg = blueSum / pixelCount

    return Color.rgb(redAvg, greenAvg, blueAvg)
 }
  ```
  and seems to be working. Should we replace old impl. with this new one ?

Reply:
weird, that should be fine. otherwise the person reporting the issue would have run into the same thing right? maybe your Android emulator is configured with too little memory? no idea tbh but I would take this much more seriously if you were running into it on a real device.

Reply:
here is my emulator config. maybe you can mess with the vm heap size and see if that resolves your issue? i think it was 256mb by default for me - so I assume that's a realistic number for real devices nowadays?
 
![image](https://user-images.githubusercontent.com/24393922/230724685-9367ad40-e7f7-4376-816d-a225573e03e7.png)

Reply:
Yeah, my bad i had strange values here 😅
But i made benchmarks for ram usage and version without `IntArray` use ~60-70 mb ram less with same color as result
What do you think ?

Reply:
No worries!

> But i made benchmarks for ram usage and version without IntArray use ~60-70 mb ram less with same color as result

Same functionality with less memory? Go for it 🤩

Reply:
Btw I'm quite the newbie at Kotlin especially when it comes to Async - a couple comments:

https://github.com/osamaqarem/react-native-image-colors/blob/20d4e869e670d42b8a3ef45d4f08492fe66a4f5a/android/src/main/java/com/reactnativeimagecolors/ImageColorsModule.kt#L144-L146

- In the above code and in other places, what I understand is that we're switching to the main thread before resolving the promise. Is there a reason that's necessary instead of just calling `promise.resolve`?

- An improvement: since this coroutine is already off of the expo async queue and the UI thread, there is no reason to switch to another thread again when generating the palette using the async signature below: 

https://github.com/osamaqarem/react-native-image-colors/blob/20d4e869e670d42b8a3ef45d4f08492fe66a4f5a/android/src/main/java/com/reactnativeimagecolors/ImageColorsModule.kt#L131-L134

instead I think we should be using the synchronous version `val palette = paletteBuilder.generate()` as we are not blocking other invocations of `getColors()`. Let me know if I'm off the mark.

Reply:
https://github.com/osamaqarem/react-native-image-colors/blob/20d4e869e670d42b8a3ef45d4f08492fe66a4f5a/android/src/main/java/com/reactnativeimagecolors/ImageColorsModule.kt#L144-L146
I am not sure what is exactly a reason but this is used in [expo documentation](https://docs.expo.dev/modules/module-api/). I guess react native needs to resolve promises on main thread for UI updates

Reply:
hey, I did some profiling on the new `calculateAverageColor` function and turns it out its significantly slower 🫣

```kotlin
val start = System.nanoTime();
result["average"] = getHex(calculateAverageColor(image, config.pixelSpacing))
val end = System.nanoTime();
val diff = end - start;
val diffSeconds = diff / 1_000_000_000.0;

val start1 = System.nanoTime();
result["average"] = getHex(calculateAverageColor2(image, config.pixelSpacing))
val end1 = System.nanoTime();
val diff1 = end1 - start1;
val diffSeconds1 = diff1 / 1_000_000_000.0;

Log.d("tag", (diffSeconds).toString() + " vs " + (diffSeconds1).toString())
```

few runs:

```
0.083448541 vs 3.051387585
0.098386666 vs 3.011872001
0.081705208 vs 2.96626996
0.068948959 vs 2.973894793
```

Tried to look into it a bit and I think calling `bitmap.getPixels(...)` once and keeping it in memory is much more performant than calling `bitmap.getPixel(...)` multiple times. I think the current performance is too slow and we should maybe revert.


----

> I am not sure what is exactly a reason but this is used in [expo documentation](https://docs.expo.dev/modules/module-api/). I guess react native needs to resolve promises on main thread for UI updates

Yup we absolutely need to be on the UI/main thread for UI updates but we aren't updating any UI here. Just moving some values between different threads.

JS (React Native) actually lives on a background thread. And by resolving the promise we are basically handing off the value from whatever thread we're on to the JS thread.

https://reactnative.dev/architecture/threading-model

![image](https://user-images.githubusercontent.com/24393922/230752049-44427f04-8a61-48a7-9b74-20656189cffe.png)


I gave it a quick test and it works without issues when resolving from the background thread. I also checked how Expo themselves tend to use the AsyncFunction API and it seems they don't do that pattern too.

https://github.com/search?q=repo:expo/expo+AsyncFunction(+language:Kotlin+promise.resolve&type=code&l=Kotlin

I think the AsyncFunction example in the docs is referring to the second bullet point. But I can see how it could be mistaken for a general example. It could be improved:

![image](https://user-images.githubusercontent.com/24393922/230751814-38beb341-ac53-456e-ba0d-066e48f76ec7.png)

Reply:
Btw another topic: I think we need to clean up properly when the module is destroyed by cancelling whatever is running.

https://docs.expo.dev/modules/module-api/#ondestroy

Found an example from an expo package:

```
OnDestroy {
  try {
    service.cancel(ModuleDestroyedException())
  } catch (e: IllegalStateException) {
    Log.e(TAG, "The scope does not have a job in it")
  }
}
```

https://github.com/expo/expo/blob/79310320c31b0cf067ac7637366b831829357983/packages/expo-video-thumbnails/android/src/main/java/expo/modules/videothumbnails/VideoThumbnailsModule.kt#L63-L69